### PR TITLE
feat: Qdrant destination support

### PR DESCRIPTION
**Pull request by Anush008**
### Description
This PR intends to add [Qdrant ](https://qdrant.tech/) as a destination to load data into using DLT.

### Documentation Preview
https://deploy-preview-724--dlt-hub-docs.netlify.app/docs/dlt-ecosystem/destinations/qdrant

### Additional Context
- The easiest way to try out this implementation is by running a local Qdrant instance using the steps mentioned [here](https://qdrant.tech/documentation/quick-start/#download-and-run). No `.dlt` configurations required, as the defaults will connect to a local instance running on port `6333`.

- Qdrant also offers an in-memory instance, which can be used by setting `credentials.location` to `:memory:`. However, it does not support multiple-writers(tests will fail) and is not recommended for production.

- Qdrant has no concrete schema, naming restrictions for data stored as payload. Any arbitrary data without any schema/property constraints can be stored.

- The [qdrant client library](https://github.com/qdrant/qdrant-client/) has built-in support for generating embeddings locally using [FastEmbed](https://github.com/qdrant/fastembed/tree/main/fastembed). When running a pipeline with Qdrant as the destination for the first time, an embedding model will be downloaded and cached locally.

- For the test action, `DESTINATION__QDRANT__CREDENTIALS__LOCATION`, `DESTINATION__QDRANT__CREDENTIALS__API_KEY` must be set in the repository secrets. A free hosted instance is available at https://cloud.qdrant.io/.

### CODE REVIEW

**Comment by Anush008** [comment]
##  NOTE: There's version conflict. 
The DLT project has Python pinned at `">=3.8.1,<4.0"`, which is incompatible with the [Qdrant client library](https://github.com/qdrant/qdrant-client). I had to set it to `">=3.8.1,<3.13"`. No breaking changes.

If this is acceptable, I'll commit the lock-file with a large diff.

**Comment by Anush008** [comment]
##  NOTE: There's a version conflict. 
The DLT project has Python pinned at `">=3.8.1,<4.0"`, which is incompatible with the [Qdrant client library](https://github.com/qdrant/qdrant-client). I had to set it to `">=3.8.1,<3.13"`. No breaking changes.

If this is acceptable, I'll commit the lock-file with a large diff.

**Comment by burnash** [comment]
Hey @Anush008, thank you for taking the time to contribute to dlt. I've started reviewing the PR and will get back to you with feedback or any necessary changes shortly.

File: dlt/destinations/qdrant/qdrant_adapter.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+            Can be a single column name as a string or a list of column names.
+
+    Returns:
+        DltResource: A resource with applied qdrant-specific hints.
+
+    Raises:
+        ValueError: If input for `embed` invalid or empty.
+
+    Examples:
+        >>> data = [{"name": "Anush", "description": "Integrations Hacker"}]
```
Comment (burnash) [inline dlt/destinations/qdrant/qdrant_adapter.py:30]: 😆

File: tests/load/qdrant/test_pipeline.py [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+from dlt.destinations.qdrant.qdrant_adapter import qdrant_adapter, VECTORIZE_HINT
+from dlt.destinations.qdrant.qdrant_client import QdrantClient
+from tests.pipeline.utils import assert_load_info
+from tests.load.qdrant.utils import drop_active_pipeline_data, assert_collection
+
+@pytest.fixture(autouse=True)
+def drop_qdrant_data() -> Iterator[None]:
+    yield
+    drop_active_pipeline_data()
+
```
Comment (burnash) [inline tests/load/qdrant/test_pipeline.py:17]:
```suggestion


```

File: tests/load/qdrant/test_pipeline.py [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+    client: QdrantClient
+    with pipeline.destination_client() as client:  # type: ignore[assignment]
+        # check if we can get a stored schema and state
+        schema = client.get_stored_schema()
+        print("Print dataset name", client.dataset_name)
+        assert schema
+        state = client.get_stored_state("test_pipeline_append")
+        assert state
+
```
Comment (burnash) [inline tests/load/qdrant/test_pipeline.py:72]:
```suggestion


```

File: tests/load/qdrant/test_pipeline.py [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+    qdrant_adapter(
+        some_data,
+        embed=["content"],
+    )
+
+    pipeline = dlt.pipeline(
+        pipeline_name="test_pipeline_append",
+        destination="qdrant",
+        dataset_name="TestPipelineAppendDataset" + uniq_id(),
+     )
```
Comment (burnash) [inline tests/load/qdrant/test_pipeline.py:90]:
```suggestion
    )
```

File: dlt/destinations/qdrant/qdrant_adapter.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
+from typing import Any
+
+from dlt.common.schema.typing import TColumnNames, TTableSchemaColumns
+from dlt.extract.decorators import resource as make_resource
+from dlt.extract.source import DltResource
+
+VECTORIZE_HINT = "x-embedding-field"
```
Comment (burnash) [inline dlt/destinations/qdrant/qdrant_adapter.py:7]: To make it clear where the hint is coming from I suggest to use a different naming: `x-qdrant-embed`.

**Review by burnash — CHANGES_REQUESTED** [review]
Hey @Anush008, once again thank you for your contribution. I've reviewed your pull request and everything looks solid for the most part. Great work! However, I noticed a few minor things that could be addressed for further improvement.

File: dlt/destinations/qdrant/qdrant_adapter.py (hunk 5 above)
Comment (Anush008) [reply dlt/destinations/qdrant/qdrant_adapter.py]: https://github.com/dlt-hub/dlt/pull/724/commits/3c25f71a36f40d7f44d737cac140afef1e1571b3

File: docs/website/docs/dlt-ecosystem/destinations/qdrant.md [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+
+```toml
+[destination.qdrant.credentials]
+location = "https://your-qdrant-url"
+api_key = "your-qdrant-api-key"
+```
+
+In this setup guide, we are using the [Qdrant Cloud](https://cloud.qdrant.io/) to get a hosted Qdrant instance and the [FastEmbed](https://github.com/qdrant/fastembed) package that is built into the [Qdrant client library](https://github.com/qdrant/qdrant-client) for generating embeddings.
+
+The `location` will default to **http://localhost:6333** and `api_key` is not defined - which are the defaults for a [local Qdrant instance](https://qdrant.tech/documentation/quick-start/#download-and-run).
```
[the remark this answers is not in the document]
Comment (Anush008) [reply docs/website/docs/dlt-ecosystem/destinations/qdrant.md]: The sentence was not quite right. What it meant was that the `qdrant_client` library defaults to `http://localhost:6333` with no API key if no config options are provided.

Updated the copy here.
https://github.com/dlt-hub/dlt/pull/724/commits/3c25f71a36f40d7f44d737cac140afef1e1571b3

File: dlt/destinations/qdrant/qdrant_client.py [hunk 7]
```diff
@@ -0,0 +1,408 @@
...
+                if self._collection_exists(qualified_table_name):
+                    continue
+
+                self.db_client.delete_collection(qualified_table_name)
+                self._create_collection(full_collection_name=qualified_table_name)
+
+    def is_storage_initialized(self) -> bool:
+        try:
+            self.db_client.get_collection(self.sentinel_collection)
+        except Exception:
```
[the remark this answers is not in the document]
Comment (Anush008) [reply dlt/destinations/qdrant/qdrant_client.py]: https://github.com/dlt-hub/dlt/pull/724/commits/3c25f71a36f40d7f44d737cac140afef1e1571b3

File: docs/website/docs/dlt-ecosystem/destinations/qdrant.md [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+
+In this setup guide, we are using the [Qdrant Cloud](https://cloud.qdrant.io/) to get a hosted Qdrant instance and the [FastEmbed](https://github.com/qdrant/fastembed) package that is built into the [Qdrant client library](https://github.com/qdrant/qdrant-client) for generating embeddings.
+
+If no configuration options are provided, the default fallback will be `http://localhost:6333` with no API key.
+
+3. Define the source of the data. For starters, let's load some data from a simple data structure:
+
+```python
+import dlt
+from dlt.destinations.qdrant import qdrant_adapter
```
Comment (burnash) [inline docs/website/docs/dlt-ecosystem/destinations/qdrant.md:36]: I tested the example from the docs and it errors with this:
```
$ python qdrant_pipeline.py 
Traceback (most recent call last):
  File "qdrant_pipeline.py", line 26, in <module>
    qdrant_adapter(
TypeError: 'module' object is not callable
```

Where `qdrant_pipeline.py` is:

```python
import dlt
from dlt.destinations.qdrant import qdrant_adapter

movies = [
    {
        "title": "Blade Runner",
        "year": 1982,
    },
    {
        "title": "Ghost in the Shell",
        "year": 1995,
    },
    {
        "title": "The Matrix",
        "year": 1999,
    }
]

pipeline = dlt.pipeline(
    pipeline_name="movies",
    destination="qdrant",
    dataset_name="MoviesDataset",
)

info = pipeline.run(
    qdrant_adapter(
        movies,
        embed="title",
    )
)

print(info)
```

**Review by burnash — CHANGES_REQUESTED** [review]

File: docs/website/docs/dlt-ecosystem/destinations/qdrant.md [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+
+5. Run the pipeline:
+
+```python
+info = pipeline.run(
+    qdrant_adapter(
+        movies,
+        embed="title",
+    )
+)
```
Comment (burnash) [inline docs/website/docs/dlt-ecosystem/destinations/qdrant.md:72]: Another problem with this sample is when I run it I get this error, even on the first run:
```
Traceback (most recent call last):
  File "/Users/b/dlthub/dlt/dlt/pipeline/pipeline.py", line 371, in load
    runner.run_pool(load.config, load)
  File "/Users/b/dlthub/dlt/dlt/common/runners/pool_runner.py", line 80, in run_pool
    while _run_func():
  File "/Users/b/dlthub/dlt/dlt/common/runners/pool_runner.py", line 73, in _run_func
    run_metrics = run_f.run(cast(TExecutor, pool))
  File "/Users/b/dlthub/dlt/dlt/load/load.py", line 405, in run
    self.load_single_package(load_id, schema)
  File "/Users/b/dlthub/dlt/dlt/load/load.py", line 325, in load_single_package
    applied_update = self._init_client(job_client, schema, expected_update, load_id, job_client.should_truncate_table_before_load, job_client.should_load_data_to_staging_dataset if isinstance(job_client, WithStagingDataset) else None)
  File "/Users/b/dlthub/dlt/dlt/load/load.py", line 307, in _init_client
    applied_update = self._init_dataset_and_update_schema(job_client, expected_update, tables_with_jobs | dlt_tables, truncate_tables)
  File "/Users/b/dlthub/dlt/dlt/load/load.py", line 296, in _init_dataset_and_update_schema
    job_client.initialize_storage(truncate_tables=truncate_tables)
  File "/Users/b/dlthub/dlt/dlt/destinations/qdrant/qdrant_client.py", line 230, in initialize_storage
    self._create_sentinel_collection()
  File "/Users/b/dlthub/dlt/dlt/destinations/qdrant/qdrant_client.py", line 246, in _create_sentinel_collection
    self._create_collection(full_collection_name=self.sentinel_collection)
  File "/Users/b/dlthub/dlt/dlt/destinations/qdrant/qdrant_client.py", line 186, in _create_collection
    self.db_client.create_collection(
  File "/Users/b/dlthub/dlt/.venv/lib/python3.8/site-packages/qdrant_client/qdrant_client.py", line 1410, in create_collection
    return self._client.create_collection(
  File "/Users/b/dlthub/dlt/.venv/lib/python3.8/site-packages/qdrant_client/qdrant_remote.py", line 1800, in create_collection
    result: Optional[bool] = self.http.collections_api.create_collection(
  File "/Users/b/dlthub/dlt/.venv/lib/python3.8/site-packages/qdrant_client/http/api/collections_api.py", line 1024, in create_collection
    return self._build_for_create_collection(
  File "/Users/b/dlthub/dlt/.venv/lib/python3.8/site-packages/qdrant_client/http/api/collections_api.py", line 82, in _build_for_create_collection
    return self.api_client.request(
  File "/Users/b/dlthub/dlt/.venv/lib/python3.8/site-packages/qdrant_client/http/api_client.py", line 74, in request
    return self.send(request, type_)
  File "/Users/b/dlthub/dlt/.venv/lib/python3.8/site-packages/qdrant_client/http/api_client.py", line 97, in send
    raise UnexpectedResponse.for_response(response)
qdrant_client.http.exceptions.UnexpectedResponse: Unexpected Response: 400 (Bad Request)
Raw response content:
b'{"status":{"error":"Wrong input: Collection `movies_dataset` already exists!"},"time":0.000015585}'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "qdrant_pipeline.py", line 25, in <module>
    info = pipeline.run(
  File "/Users/b/dlthub/dlt/dlt/pipeline/pipeline.py", line 109, in _wrap
    step_info = f(self, *args, **kwargs)
  File "/Users/b/dlthub/dlt/dlt/pipeline/pipeline.py", line 143, in _wrap
    return f(self, *args, **kwargs)
  File "/Users/b/dlthub/dlt/dlt/pipeline/pipeline.py", line 465, in run
    return self.load(destination, dataset_name, credentials=credentials)
  File "/Users/b/dlthub/dlt/dlt/pipeline/pipeline.py", line 109, in _wrap
    step_info = f(self, *args, **kwargs)
  File "/Users/b/dlthub/dlt/dlt/pipeline/pipeline.py", line 83, in _wrap
    rv = f(self, *args, **kwargs)
  File "/Users/b/dlthub/dlt/dlt/pipeline/pipeline.py", line 69, in _wrap
    return f(self, *args, **kwargs)
  File "/Users/b/dlthub/dlt/dlt/pipeline/pipeline.py", line 143, in _wrap
    return f(self, *args, **kwargs)
  File "/Users/b/dlthub/dlt/dlt/pipeline/pipeline.py", line 376, in load
    raise PipelineStepFailed(self, "load", l_ex, self._get_load_info(load)) from l_ex
dlt.pipeline.exceptions.PipelineStepFailed: Pipeline execution failed at stage load with exception:

<class 'qdrant_client.http.exceptions.UnexpectedResponse'>
Unexpected Response: 400 (Bad Request)
Raw response content:
b'{"status":{"error":"Wrong input: Collection `movies_dataset` already exists!"},"time":0.000015585}'
```

**Review by burnash — CHANGES_REQUESTED** [review]

File: dlt/destinations/qdrant/qdrant_client.py [hunk 10]
```diff
@@ -0,0 +1,404 @@
...
+                qualified_table_name = self._make_qualified_collection_name(
+                    table_name=table_name)
+                if self._collection_exists(qualified_table_name):
+                    continue
+
+                self.db_client.delete_collection(qualified_table_name)
+                self._create_collection(full_collection_name=qualified_table_name)
+
+    def is_storage_initialized(self) -> bool:
+        self._collection_exists(self.sentinel_collection)
```
Comment (burnash) [inline dlt/destinations/qdrant/qdrant_client.py:242]:
```suggestion
        return self._collection_exists(self.sentinel_collection)
```
I think that is the cause of the failed pipeline I mentioned.

**Review by burnash — CHANGES_REQUESTED** [review]

File: docs/website/docs/dlt-ecosystem/destinations/qdrant.md (hunk 8 above)
Comment (Anush008) [reply docs/website/docs/dlt-ecosystem/destinations/qdrant.md:36]: https://github.com/dlt-hub/dlt/pull/724/commits/7980f386fac41053d49d7a6c351b64ef38499dba

File: docs/website/docs/dlt-ecosystem/destinations/qdrant.md (hunk 9 above)
Comment (Anush008) [reply docs/website/docs/dlt-ecosystem/destinations/qdrant.md:72]: https://github.com/dlt-hub/dlt/pull/724/commits/269d2d40b5b587ac240acd4dbb8ad37ae9bfa159

**Review by burnash — CHANGES_REQUESTED** [review]

File: docs/website/docs/dlt-ecosystem/destinations/qdrant.md [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+```
+
```
[the remark this answers is not in the document]
Comment (Anush008) [reply docs/website/docs/dlt-ecosystem/destinations/qdrant.md:74]: This seems like there were problems earlier when setting the schema.
Could you delete the collection at http://localhost:6333/dashboard? Or use a different dataset name?

File: docs/website/docs/dlt-ecosystem/destinations/qdrant.md [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+5. Run the pipeline:
+
+```python
+info = pipeline.run(
+    qdrant_adapter(
+        movies,
+        embed="title",
+    )
+)
+```
```
[the remark this answers is not in the document]
Comment (Anush008) [reply docs/website/docs/dlt-ecosystem/destinations/qdrant.md:73]: I think I got the cause here. Let me push a fix.

[the remark this answers is not in the document]
Comment (Anush008) [reply docs/website/docs/dlt-ecosystem/destinations/qdrant.md:73]: https://github.com/Anush008/dlt/commit/bcda2cb1413f7840facac4bc758362d96b5fad04. Should be fixed now.

File: docs/website/docs/dlt-ecosystem/destinations/qdrant.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (burnash) [reply docs/website/docs/dlt-ecosystem/destinations/qdrant.md:74]: @Anush008 The error persists both when I used a clean database and tried a different dataset name. Notably the collections got created but they are all empty except `movies_dataset__dlt_version`:

```
[2023-11-05T17:51:41.986+0000] {_client.py:1013} INFO - HTTP Request: POST https://xxxxx.europe-west3-0.gcp.cloud.qdrant.io:6333/collections/movies_dataset2__dlt_pipeline_state/points/scroll "HTTP/1.1 404 Not Found"
[2023-11-05T17:51:42.916+0000] {_client.py:1013} INFO - HTTP Request: GET https://xxxxx.europe-west3-0.gcp.cloud.qdrant.io:6333/collections/movies_dataset2_movies_dataset2 "HTTP/1.1 404 Not Found"
[2023-11-05T17:51:43.099+0000] {_client.py:1013} INFO - HTTP Request: PUT https://xxxxx.europe-west3-0.gcp.cloud.qdrant.io:6333/collections/movies_dataset2 "HTTP/1.1 200 OK"
[2023-11-05T17:51:43.146+0000] {_client.py:1013} INFO - HTTP Request: POST https://xxxxx.europe-west3-0.gcp.cloud.qdrant.io:6333/collections/movies_dataset2__dlt_version/points/scroll "HTTP/1.1 404 Not Found"
[2023-11-05T17:51:43.189+0000] {_client.py:1013} INFO - HTTP Request: GET https://xxxxx.europe-west3-0.gcp.cloud.qdrant.io:6333/collections/movies_dataset2_content "HTTP/1.1 404 Not Found"
[2023-11-05T17:51:43.351+0000] {_client.py:1013} INFO - HTTP Request: PUT https://xxxxx.europe-west3-0.gcp.cloud.qdrant.io:6333/collections/movies_dataset2_content "HTTP/1.1 200 OK"
[2023-11-05T17:51:43.396+0000] {_client.py:1013} INFO - HTTP Request: GET https://xxxxx.europe-west3-0.gcp.cloud.qdrant.io:6333/collections/movies_dataset2__dlt_pipeline_state "HTTP/1.1 404 Not Found"
[2023-11-05T17:51:43.595+0000] {_client.py:1013} INFO - HTTP Request: PUT https://xxxxx.europe-west3-0.gcp.cloud.qdrant.io:6333/collections/movies_dataset2__dlt_pipeline_state "HTTP/1.1 200 OK"
[2023-11-05T17:51:43.635+0000] {_client.py:1013} INFO - HTTP Request: GET https://xxxxx.europe-west3-0.gcp.cloud.qdrant.io:6333/collections/movies_dataset2__dlt_version "HTTP/1.1 404 Not Found"
[2023-11-05T17:51:43.828+0000] {_client.py:1013} INFO - HTTP Request: PUT https://xxxxx.europe-west3-0.gcp.cloud.qdrant.io:6333/collections/movies_dataset2__dlt_version "HTTP/1.1 200 OK"
[2023-11-05T17:51:43.871+0000] {_client.py:1013} INFO - HTTP Request: GET https://xxxxx.europe-west3-0.gcp.cloud.qdrant.io:6333/collections/movies_dataset2__dlt_loads "HTTP/1.1 404 Not Found"
[2023-11-05T17:51:44.036+0000] {_client.py:1013} INFO - HTTP Request: PUT https://xxxxx.europe-west3-0.gcp.cloud.qdrant.io:6333/collections/movies_dataset2__dlt_loads "HTTP/1.1 200 OK"
[2023-11-05T17:51:44.085+0000] {_client.py:1013} INFO - HTTP Request: PUT https://xxxxx.europe-west3-0.gcp.cloud.qdrant.io:6333/collections/movies_dataset2__dlt_version/points?wait=true "HTTP/1.1 200 OK"
[2023-11-05T17:51:44.125+0000] {_client.py:1013} INFO - HTTP Request: GET https://xxxxx.europe-west3-0.gcp.cloud.qdrant.io:6333/collections/movies_dataset2_movies_dataset2 "HTTP/1.1 404 Not Found"
[2023-11-05T17:51:44.165+0000] {_client.py:1013} INFO - HTTP Request: PUT https://xxxxx.europe-west3-0.gcp.cloud.qdrant.io:6333/collections/movies_dataset2 "HTTP/1.1 400 Bad Request"
Traceback (most recent call last):
  File "/Users/b/dlthub/dlt/dlt/pipeline/pipeline.py", line 371, in load
    runner.run_pool(load.config, load)
  File "/Users/b/dlthub/dlt/dlt/common/runners/pool_runner.py", line 80, in run_pool
    while _run_func():
  File "/Users/b/dlthub/dlt/dlt/common/runners/pool_runner.py", line 73, in _run_func
    run_metrics = run_f.run(cast(TExecutor, pool))
  File "/Users/b/dlthub/dlt/dlt/load/load.py", line 405, in run
    self.load_single_package(load_id, schema)
  File "/Users/b/dlthub/dlt/dlt/load/load.py", line 325, in load_single_package
    applied_update = self._init_client(job_client, schema, expected_update, load_id, job_client.should_truncate_table_before_load, job_client.should_load_data_to_staging_dataset if isinstance(job_client, WithStagingDataset) else None)
  File "/Users/b/dlthub/dlt/dlt/load/load.py", line 307, in _init_client
    applied_update = self._init_dataset_and_update_schema(job_client, expected_update, tables_with_jobs | dlt_tables, truncate_tables)
  File "/Users/b/dlthub/dlt/dlt/load/load.py", line 296, in _init_dataset_and_update_schema
    job_client.initialize_storage(truncate_tables=truncate_tables)
  File "/Users/b/dlthub/dlt/dlt/destinations/qdrant/qdrant_client.py", line 230, in initialize_storage
    self._create_sentinel_collection()
  File "/Users/b/dlthub/dlt/dlt/destinations/qdrant/qdrant_client.py", line 246, in _create_sentinel_collection
    self._create_collection(full_collection_name=self.sentinel_collection)
  File "/Users/b/dlthub/dlt/dlt/destinations/qdrant/qdrant_client.py", line 186, in _create_collection
    self.db_client.create_collection(
  File "/Users/b/dlthub/dlt/.venv/lib/python3.8/site-packages/qdrant_client/qdrant_client.py", line 1410, in create_collection
    return self._client.create_collection(
  File "/Users/b/dlthub/dlt/.venv/lib/python3.8/site-packages/qdrant_client/qdrant_remote.py", line 1800, in create_collection
    result: Optional[bool] = self.http.collections_api.create_collection(
  File "/Users/b/dlthub/dlt/.venv/lib/python3.8/site-packages/qdrant_client/http/api/collections_api.py", line 1024, in create_collection
    return self._build_for_create_collection(
  File "/Users/b/dlthub/dlt/.venv/lib/python3.8/site-packages/qdrant_client/http/api/collections_api.py", line 82, in _build_for_create_collection
    return self.api_client.request(
  File "/Users/b/dlthub/dlt/.venv/lib/python3.8/site-packages/qdrant_client/http/api_client.py", line 74, in request
    return self.send(request, type_)
  File "/Users/b/dlthub/dlt/.venv/lib/python3.8/site-packages/qdrant_client/http/api_client.py", line 97, in send
    raise UnexpectedResponse.for_response(response)
qdrant_client.http.exceptions.UnexpectedResponse: Unexpected Response: 400 (Bad Request)
Raw response content:
b'{"status":{"error":"Wrong input: Collection `movies_dataset2` already exists!"},"time":0.00001558}'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "qdrant_pipeline.py", line 25, in <module>
    info = pipeline.run(
  File "/Users/b/dlthub/dlt/dlt/pipeline/pipeline.py", line 109, in _wrap
    step_info = f(self, *args, **kwargs)
  File "/Users/b/dlthub/dlt/dlt/pipeline/pipeline.py", line 143, in _wrap
    return f(self, *args, **kwargs)
  File "/Users/b/dlthub/dlt/dlt/pipeline/pipeline.py", line 471, in run
    return self.load(destination, dataset_name, credentials=credentials)
  File "/Users/b/dlthub/dlt/dlt/pipeline/pipeline.py", line 109, in _wrap
    step_info = f(self, *args, **kwargs)
  File "/Users/b/dlthub/dlt/dlt/pipeline/pipeline.py", line 83, in _wrap
    rv = f(self, *args, **kwargs)
  File "/Users/b/dlthub/dlt/dlt/pipeline/pipeline.py", line 69, in _wrap
    return f(self, *args, **kwargs)
  File "/Users/b/dlthub/dlt/dlt/pipeline/pipeline.py", line 143, in _wrap
    return f(self, *args, **kwargs)
  File "/Users/b/dlthub/dlt/dlt/pipeline/pipeline.py", line 376, in load
    raise PipelineStepFailed(self, "load", l_ex, self._get_load_info(load)) from l_ex
dlt.pipeline.exceptions.PipelineStepFailed: Pipeline execution failed at stage load with exception:

<class 'qdrant_client.http.exceptions.UnexpectedResponse'>
Unexpected Response: 400 (Bad Request)
Raw response content:
b'{"status":{"error":"Wrong input: Collection `movies_dataset2` already exists!"},"time":0.00001558}'
```

File: docs/website/docs/dlt-ecosystem/destinations/qdrant.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (burnash) [reply docs/website/docs/dlt-ecosystem/destinations/qdrant.md:73]: Yes, it works now with a local setup.

File: docs/website/docs/dlt-ecosystem/destinations/qdrant.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (Anush008) [reply docs/website/docs/dlt-ecosystem/destinations/qdrant.md:74]: > https://github.com/dlt-hub/dlt/pull/724#discussion_r1382620025.

Is the issue resolved?

File: docs/website/docs/dlt-ecosystem/destinations/qdrant.md [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+
+6. Check the results:
+
+```python
+print(info)
+```
+
+The data is now loaded into Qdrant.
```
Comment (burnash) [inline docs/website/docs/dlt-ecosystem/destinations/qdrant.md:81]: Another question regarding this example. After loading the data, how to query it with `client.query`? As in:

```python
# pip install qdrant-client[fastembed]

from qdrant_client import QdrantClient

client = QdrantClient()

result = client.query(
    collection_name="content",
    query_text="Neo",
)
print(result)
```

When I run the above I get:
```
Traceback (most recent call last):
  File "qdrant_query.py", line 23, in <module>
    search_result = client.query(
  File "/Users/b/dlthub/dlt/.venv/lib/python3.8/site-packages/qdrant_client/qdrant_fastembed.py", line 333, in query
    self.search(
  File "/Users/b/dlthub/dlt/.venv/lib/python3.8/site-packages/qdrant_client/qdrant_client.py", line 315, in search
    return self._client.search(
  File "/Users/b/dlthub/dlt/.venv/lib/python3.8/site-packages/qdrant_client/qdrant_remote.py", line 454, in search
    search_result = self.http.points_api.search_points(
  File "/Users/b/dlthub/dlt/.venv/lib/python3.8/site-packages/qdrant_client/http/api/points_api.py", line 1197, in search_points
    return self._build_for_search_points(
  File "/Users/b/dlthub/dlt/.venv/lib/python3.8/site-packages/qdrant_client/http/api/points_api.py", line 536, in _build_for_search_points
    return self.api_client.request(
  File "/Users/b/dlthub/dlt/.venv/lib/python3.8/site-packages/qdrant_client/http/api_client.py", line 74, in request
    return self.send(request, type_)
  File "/Users/b/dlthub/dlt/.venv/lib/python3.8/site-packages/qdrant_client/http/api_client.py", line 97, in send
    raise UnexpectedResponse.for_response(response)
qdrant_client.http.exceptions.UnexpectedResponse: Unexpected Response: 400 (Bad Request)
Raw response content:
b'{"status":{"error":"Wrong input: Vector params for fast-bge-small-en are not specified in config"},"time":0.002777542}'
```

At the same time `client.search()` works as expected:

```python
query_vector = [.0] * 384

result = client.search(
    collection_name="content",
    query_vector=query_vector,
    limit=3,
)
print(result)

>>> [ScoredPoint(id='00b14076-69bf-536b-aa97-2a98435e4fcc', version=2, score=0.0, payload={'_dlt_id': 'CbMonMJoPDlfKg', '_dlt_load_id': '1699207571.018365', 'title': 'The Matrix', 'year': 1999}, vector=None), ScoredPoint(id='f15418fe-ea92-5b83-98b0-4829cd79b322', version=2, score=0.0, payload={'_dlt_id': 'wgwJEYdunRhldg', '_dlt_load_id': '1699207571.018365', 'title': 'Ghost in the Shell', 'year': 1995}, vector=None), ScoredPoint(id='f3fdeb51-103f-54e1-9fb8-19b4d2abfd8c', version=2, score=0.0, payload={'_dlt_id': 'v4uJ+VeY3gJ83g', '_dlt_load_id': '1699207571.018365', 'title': 'Blade Runner', 'year': 1982}, vector=None)]
```

**Review by burnash — CHANGES_REQUESTED** [review]

File: docs/website/docs/dlt-ecosystem/destinations/qdrant.md (hunk 13 above)
Comment (Anush008) [reply docs/website/docs/dlt-ecosystem/destinations/qdrant.md:81]: I'll have to update the insertion for this.
The wquery()` method expects a named vector. Where the name is the the name of the model.

File: docs/website/docs/dlt-ecosystem/destinations/qdrant.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (burnash) [reply docs/website/docs/dlt-ecosystem/destinations/qdrant.md:74]: Yes, the credentials error is fixed.

[the remark this answers is not in the document]
Comment (Anush008) [reply docs/website/docs/dlt-ecosystem/destinations/qdrant.md:74]: This one should be resolved now.
https://github.com/dlt-hub/dlt/pull/724/commits/86f5b979fbbf0966bbc7820c5f70c10b48b3721a

File: docs/website/docs/dlt-ecosystem/destinations/qdrant.md (hunk 13 above)
Comment (Anush008) [reply docs/website/docs/dlt-ecosystem/destinations/qdrant.md:81]: https://github.com/dlt-hub/dlt/pull/724/commits/86f5b979fbbf0966bbc7820c5f70c10b48b3721a.
You can now perform `query()`. 
And `search()` by specifying the vector name using `<QdrantClient>.get_vector_field_name()`.

**Comment by burnash** [comment]
@Anush008 the CI is failing and I need to take a closer look why

**Comment by Anush008** [comment]
> @Anush008 the CI is failing and I need to take a closer look why

Maybe because of the lockfile I told you about.

**Comment by Anush008** [comment]
@burnash, just pushed the lockfile.

File: .github/workflows/test_destination_qdrant.yml [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+jobs:
+  get_docs_changes:
+    uses: ./.github/workflows/get_docs_changes.yml
+    if: ${{ !github.event.pull_request.head.repo.fork }}
+
+  run_loader:
+    name: Tests Qdrant loader
+    needs: get_docs_changes
+    if: !always()
+    # if: needs.get_docs_changes.outputs.changes_outside_docs == 'true'
```
Comment (burnash) [inline .github/workflows/test_destination_qdrant.yml:29]:
```suggestion
    if: needs.get_docs_changes.outputs.changes_outside_docs == 'true'
```

**Review by burnash — CHANGES_REQUESTED** [review]

**Comment by burnash** [comment]
@Anush008 could you please apply suggested changes (that'd enable the ci) and rebase the branch on devel?

**Comment by Anush008** [comment]
@burnash, should be good now.

File: dlt/common/pendulum.py [hunk 15]
```diff
@@ -2,7 +2,7 @@
 import pendulum  # noqa: I251
 
 # force UTC as the local timezone to prevent local dates to be written to dbs
-pendulum.set_local_timezone(pendulum.timezone('UTC'))
+pendulum.set_local_timezone(pendulum.timezone('UTC')) # type: ignore
```
Comment (burnash) [inline dlt/common/pendulum.py:5]: This errors with `Unused "type: ignore" comment  [unused-ignore]` in CI
https://github.com/dlt-hub/dlt/actions/runs/6786606444/job/18447614187?pr=724

Comment (Anush008) [reply dlt/common/pendulum.py:5]: Weird that running lint on my machine works fine. Removing this ignore label errors.

```bash
make lint
./check-package.sh
poetry run mypy --config-file mypy.ini dlt tests
dlt/common/pendulum.py:5: error: Module not callable  [operator]
Found 1 error in 1 file (checked 474 source files)
make: *** [lint] Error 1
```

Comment (Anush008) [reply dlt/common/pendulum.py]: I committed the change nevertheless.

Comment (Anush008) [reply dlt/common/pendulum.py]: Passed.

**Comment by Anush008** [comment]
```bash
=========================== short test summary info ============================
FAILED examples/chess_production/code/chess-snippets.py::incremental_snippet
FAILED examples/incremental_loading/code/zendesk-snippets.py::incremental_snippet
============= 2 failed, 17 passed, 23 warnings in 76.15s (0:01:16) =============
```
@burnash, the failing doc tests are expected to be skipped for forks.
https://github.com/dlt-hub/dlt/blob/7f2f042815fd548a35aa202629b7ddb443b8399b/docs/website/docs/examples/incremental_loading/code/zendesk-snippets.py#L8-L9
https://github.com/dlt-hub/dlt/blob/7f2f042815fd548a35aa202629b7ddb443b8399b/docs/website/docs/examples/chess_production/code/chess-snippets.py#L4-L5

**Comment by burnash** [comment]
@Anush008 yes, please ignore these

**Comment by Anush008** [comment]
Great. I believe we're good with this then.

**Comment by burnash** [comment]
@Anush008, a huge thank you for your hard work and dedication. Excited to see the impact of having Qdrant on board!

**Pull request merged**